### PR TITLE
feat(developer): improve compiler messages and user interface

### PR DESCRIPTION
**Pull request by mcdurdin**
* Add unit test to verify the format of message detail strings (max line length of 80 chars)
* Add unit test to verify non-null, non-empty message strings
* Tidy up a number of messages (to pass unit tests)
* Add message details for all kmw compiler messages
* Add a clickable link for the TIKE message window for each message so that explanations and further message details are more accessible to authors (`kmc message` is already available for command-line users)
* Fixes a number of minor issues with the messages window in TIKE, mostly relating to when popup menu commands are enabled and how messages are selected.

Fixes: #13134

# User Testing

* TEST_MESSAGE_LINKS: Open a keyboard project, compile it, and attempt to click on the KM##### links in the message window that appears. Verify that the links go to a working page which relates to the selected message.
* TEST_MESSAGE_RIGHT-CLICK: Open a keyboard project, compile it. Right click on any message in the message window and verify
that selecting 'Open documentation on selected message' opens the related message documentation.

### CODE REVIEW

File: common/windows/delphi/general/Upload_Settings.pas [hunk 1]
```diff
@@ -179,4 +181,9 @@ function URLPath_PackageDownload(const PackageID, BCP47: string; IsUpdate: Boole
   Result := Format(URLPath_PackageDownload_Format, [URLEncode(PackageID), URLEncode(CKeymanVersionInfo.Tier), URLEncode(BCP47), IsUpdateInt]);
 end;
 
+function URL_KmcMessage(id: string): string;
+begin
+   Result := S_HelpKeymanCom + '/developer/'+GetMajorMinorVersionString+'/reference/messages/'+id.ToLower;
```
Comment (mcdurdin) [inline common/windows/delphi/general/Upload_Settings.pas:186]: I now realize that this violates our earlier rule of always using /go/ urls in code. Thoughts?

File: developer/src/kmc-copy/src/copier-messages.ts [hunk 2]
```diff
@@ -19,9 +19,10 @@ const SevFatal = CompilerErrorSeverity.Fatal | Namespace;
  */
 export class CopierMessages {
   static FATAL_UnexpectedException = SevFatal | 0x0001;
-  static Fatal_UnexpectedException = (o:{e: any}) => m(
+  static Fatal_UnexpectedException = (o:{e: any}) => CompilerMessageSpecWithException(
```
Comment (mcdurdin) [inline developer/src/kmc-copy/src/copier-messages.ts:22]: This was incorrect -- was giving a null error message, picked up by the new test :grin:

File: developer/src/tike/project/Keyman.Developer.UI.Project.UfrmCloneLocalProjectParameters.pas [hunk 3]
```diff
@@ -199,7 +199,7 @@ function TfrmCloneLocalProjectParameters.IsSourceProjectALexicalModelProject: Bo
 procedure TfrmCloneLocalProjectParameters.EnableControls;
 var
   e: Boolean;
-  sourceId, id: string;
```
Comment (mcdurdin) [inline developer/src/tike/project/Keyman.Developer.UI.Project.UfrmCloneLocalProjectParameters.pas:202]: unrelated, unused variable hint in the compilation that I spotted.

File: common/windows/delphi/general/Upload_Settings.pas (hunk 1 above)
Comment (mcdurdin) [reply common/windows/delphi/general/Upload_Settings.pas:186]: Note that the compiler error message details even in this PR also contain long-form URLs, so perhaps we need to revisit this rule?

File: common/windows/delphi/general/VersionInfo.pas [hunk 4]
```diff
@@ -54,6 +55,12 @@ function GetVersionString: WideString;
             IntToStr(LoWord(VersionMinor));
 end;
 
+function GetMajorMinorVersionString: string;
+begin
+  Result := IntToStr(HiWord(VersionMajor)) + '.' +  // I3310
+            IntToStr(LoWord(VersionMajor));
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/windows/delphi/general/VersionInfo.pas:61]: > Naming-wise, I find it a little confusing that `getMajorMinorVersionString` only uses `VersionMajor`,
> while `GetVersionString` uses both `VersionMajor` and `VersionMinor`.

Yeah, collision between Windows version number component naming (major high/low, minor high/low maj.maj.min.min) vs semver formats (major.minor.patch). I was using semver format for the function name. A bit squidgy.

File: developer/src/common/web/utils/src/compiler-interfaces.ts [hunk 5]
```diff
@@ -399,6 +399,19 @@ export const CompilerMessageSpec = (code: number, message: string, detail?: stri
...
+ * Remove initial whitespace from compiler detail messages, to enable
+ * indented formatting of message detail strings inside the message
+ * definitions
+ * @param event
+ * @returns dedented event detail
+ */
+export function dedentCompilerMessageDetail(event: CompilerEvent) {
+  // TODO(lowpri): dedent may be too naive -- should use least
+  // non-zero whitespace line as amount to dedent
+  return (event.detail ?? '').replace(/^[ ]+/gm, '');
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply developer/src/common/web/utils/src/compiler-interfaces.ts:412]: No, I would prefer not. We should not be using tabs in the message.ts source files for indents.

File: common/windows/delphi/general/Upload_Settings.pas (hunk 1 above)
Comment (mcdurdin) [reply common/windows/delphi/general/Upload_Settings.pas:186]: Looking at what we already have with `kmc message`, I think we should go with `https://kmn.sh/km#####` as our redirect (already baked in). It's also a redirect with intent for URLs to be permanent so if help site layout changes then it will just mean updating the redirect.

I am not so sure about URLs embedded in message details. Am considering leaving those as-is, and declaring them to be an exception to the rule, because otherwise we have to add a lot of redirects with little benefit.

**Comment by mcdurdin** [comment]
@darcywong00, have addressed review comments

File: common/windows/delphi/general/VersionInfo.pas (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/windows/delphi/general/VersionInfo.pas:61]: > Is there any impact this returns a `string` while all the other methods return `WideString`?

no impact. WideString was the norm for Unicode strings before string type changed to Unicode (quite) a few versions of Delphi ago. string is a little more performant. I decided not to changeover existing usages of WideString unless we were doing maintenance on that bit of code.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by dinakaranr** [comment]
# Test Results
I tested this issue with the attached Keyman"18.0.185-alpha-test-13156" build(08/02/2025) on Windows 10. Here I am sharing my observation.

* **TEST_MESSAGE_LINKS (Passed):**
1. Install Keyman attached the "18.0.185-alpha-test-13156" PR build.
2. Open the "gautami_devanagari" keyboard project loaded in the keyman developer.
3. Compile the keyboard by clicking the keyboard menu --> compile keyboard.
4. Verified that the compile success message appeared.
5. Click on the "KM####" link from the compiled message.
6. The link appeared on the browser.
7. Verified the message details appeared on the "help.keyman.com" website.
It works well. Thank you.

* **TEST_MESSAGE_RIGHT-CLICK (Passed):**
1. Install Keyman attached the "18.0.185-alpha-test-13156" PR build.
2. Open the "gautami_devanagari" keyboard project loaded in the keyman developer.
3. Compile the keyboard by clicking the keyboard menu --> compile keyboard.
4. Verified that the compile success message appeared.
5. Right-click on the compiled message and then select the "Open documentation on selected message" option from the menu.
6. The link appeared on the browser.
7. Verified the message details appeared on the "help.keyman.com" website.
It works well. Thank you.

File: common/windows/delphi/general/Upload_Settings.pas (hunk 1 above)
Comment (mcdurdin) [reply common/windows/delphi/general/Upload_Settings.pas]: I've made validation of hard-coded URLs in kmc messages into a future issue #13164; FYI @darcywong00

**Review by rc-swag — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.188-alpha](https://keyman.com/downloads/releases/alpha/18.0.188)
